File: c/cloud.c
#include <stdlib.h>
#include <raylib.h>
#include <stdio.h>
#include <stdlib.h>

#include "cloud.h"

#define BASE_SIZE_X 160
#define BASE_SIZE_Y 160

void parseSignature(Cloud* cloud, char* hash, int hashCount);
void generateSignature(Cloud* cloud);

Cloud createCloud(float x, float y) {
    Cloud cloud = (Cloud){
        .x = x,
        .y = y,
    };
    
    generateSignature(&cloud);
    
    return cloud;
}

Cloud createCloudWithSignature(float x, float y, char* hash, int hashCount) {
    Cloud cloud = (Cloud){
        .x = x,
        .y = y,
    };
    
    parseSignature(&cloud, hash, hashCount);
    
    return cloud;
}

void drawCloud(Cloud* cloud) {
    for (int a = 0; a < cloud->signature[0]; a++) {
        int size = cloud->signature[a * 3 + 1];
        int left = cloud->signature[a * 3 + 2];
        int top = cloud->signature[a * 3 + 3];
        
        DrawCircle((int)cloud->x + left, (int)cloud->y + top, (float)size, WHITE);
    }
}

void parseSignature(Cloud* cloud, char* hash, int hashCount) {
    char digits[6];
    int digitCount = 0;
    int partCount = 0;
    
    for (int a = 0; a < hashCount; a++) {
        if (hash[a] == ':') {
            digits[digitCount] = 0;
            digitCount = 0;
            
            cloud->signature[partCount] = atoi(digits);
            
            partCount += 1;
        }
        else {
            digits[digitCount++] = hash[a];
        }
    }
    
    digits[digitCount] = 0;
    cloud->signature[partCount] = atoi(digits);
}

void generateSignature(Cloud* cloud) {
    int numCircle = (rand() % 10) + 10;
    cloud->signature[0] = numCircle;
    
    for (int a = 0; a < numCircle; a++) {
        int size = (rand() % 50) + 15;
        int left = rand() % BASE_SIZE_X;
        int top = rand() % BASE_SIZE_Y;
        
        cloud->signature[a * 3 + 1] = size;
        cloud->signature[a * 3 + 2] = left;
        cloud->signature[a * 3 + 3] = top;
    }
}

File: go/cloud.go
package main

import (
	"math/rand"

	rl "github.com/gen2brain/raylib-go/raylib"
)

type Cloud struct {
	X float32
	Y float32
	Signature []int32
}

const baseSizeX = 160
const baseSizeY = 160

func createCloud(x float32, y float32) Cloud {
	return Cloud{
		X: x,
		Y: y,
		Signature: generateSignature(),
	}
}

func (cloud *Cloud) draw() {
	for i := range cloud.Signature[0] {
        size := cloud.Signature[i * 3 + 1]
        left := cloud.Signature[i * 3 + 2]
        top := cloud.Signature[i * 3 + 3]
        
        rl.DrawCircle(int32(cloud.X) + left, int32(cloud.Y) + top, float32(size), rl.White)
    }
}

func generateSignature() []int32 {
	res := []int32{}
	
	numCircle := rand.Intn(10) + 10
	res = append(res, int32(numCircle))
	
	for range numCircle {
		size := rand.Intn(50) + 15
		left := rand.Intn(baseSizeX)
		top := rand.Intn(baseSizeY)
		
		res = append(res, int32(size))
		res = append(res, int32(left))
		res = append(res, int32(top))
	}
	
	return res
}